service/tbc/peer_manager.go

```diff
@@ -5,129 +5,349 @@
+		good:     make(map[string]struct{}, maxPeersGood),
+		bad:      make(map[string]struct{}, maxPeersBad),
+		peers:    make(map[string]*peer, want),
+	}, nil
+}
+
+func (pm *PeerManager) String() string {
+	pm.mtx.RLock()
+	defer pm.mtx.RUnlock()
+
+	return fmt.Sprintf("connected %v good %v bad %v",
+		len(pm.peers), len(pm.good), len(pm.bad))
+}
+
+func (pm *PeerManager) seed(pctx context.Context) error {
+	log.Tracef("seed")
+	defer log.Tracef("seed exit")
+
+	// Seed
+	resolver := &net.Resolver{}
+	ctx, cancel := context.WithTimeout(pctx, 15*time.Second)
+	defer cancel()
+
+	errorsSeen := 0
+	for _, v := range pm.dnsSeeds {
+		host, port, err := net.SplitHostPort(v)
+		if err != nil {
+			log.Errorf("Failed to parse host/port: %v", err)
+			errorsSeen++
+			continue
+		}
+		ips, err := resolver.LookupIP(ctx, "ip", host)
+		if err != nil {
+			log.Errorf("lookup: %v", err)
+			errorsSeen++
+			continue
+		}
+
+		for _, ip := range ips {
+			address := net.JoinHostPort(ip.String(), port)
+			pm.seeds = append(pm.seeds, address)
+		}
+	}
+
+	if len(pm.seeds) == 0 {
+		return errors.New("could not dns seed")
 	}
+
+	return nil
 }
 
 // Stats returns peer statistics.
-func (pm *PeerManager) Stats() (int, int) {
+func (pm *PeerManager) Stats() (int, int, int) {
 	log.Tracef("PeersStats")
 	defer log.Tracef("PeersStats exit")
 
-	pm.peersMtx.RLock()
-	defer pm.peersMtx.RUnlock()
-	return len(pm.peersGood), len(pm.peersBad)
+	pm.mtx.RLock()
+	defer pm.mtx.RUnlock()
+	return len(pm.peers), len(pm.good), len(pm.bad)
 }
 
-// PeersInsert adds known peers.
-func (pm *PeerManager) PeersInsert(peers []string) error {
-	log.Tracef("PeersInsert")
-	defer log.Tracef("PeersInsert exit")
-
-	pm.peersMtx.Lock()
+func (pm *PeerManager) handleAddr(peers []string) {
 	for _, addr := range peers {
-		if _, ok := pm.peersBad[addr]; ok {
+		_, _, err := net.SplitHostPort(addr)
+		if err != nil {
+			continue
+		}
+		if _, ok := pm.peers[addr]; ok {
+			// Skip connected peers.
+			continue
+		}
+		if _, ok := pm.bad[addr]; ok {
 			// Skip bad peers.
 			continue
 		}
-		pm.peersGood[addr] = struct{}{}
+		pm.good[addr] = struct{}{}
+	}
+	log.Debugf("HandleAddr exit %v good %v bad %v",
+		len(peers), len(pm.good), len(pm.bad))
+}
+
+// HandleAddr adds peers to good list.
+func (pm *PeerManager) HandleAddr(peers []string) {
+	log.Tracef("HandleAddr %v", len(peers))
+
+	pm.mtx.Lock()
+	defer pm.mtx.Unlock()
+	pm.handleAddr(peers)
+}
+
+// Good adds peer good list.
+func (pm *PeerManager) Good(address string) error {
+	log.Tracef("Good")
+	defer log.Tracef("Good exit")
+
+	_, _, err := net.SplitHostPort(address)
+	if err != nil {
+		return err
 	}
-	allGoodPeers := len(pm.peersGood)
-	allBadPeers := len(pm.peersBad)
-	pm.peersMtx.Unlock()
 
-	log.Debugf("PeersInsert exit %v good %v bad %v",
-		len(peers), allGoodPeers, allBadPeers)
+	pm.mtx.Lock()
+	defer pm.mtx.Unlock()
+
+	// If peer is connected don't add it to good list
+	if _, ok := pm.peers[address]; ok {
```

Comment:
is this an _error_ or should it just be a no-op?

Reply:
Fixed.

Reply:
error for diagnostic.